server/src/main/java/com/google/coffeehouse/servlets/UpdateClubServlet.java

```diff
@@ -0,0 +1,197 @@
+  }
+
+  /** 
+   * Updates a Club object in the database and returns that object in JSON format.
+   * @param request the POST request that must have a valid JSON representation of the Club to be
+   *     updated, the {@code userId} of the user who wants to update the Club, as well as an
+   *     optional mask of fields that will be updated as its body. The Club to be updated must be
+   *     represented by a {@code "club"} key, whose value is an object that represents the updated
+   *     Club. Inside of this object, there will be a {@code "currentBook"} key that is associated
+   *     with an object that represents the Book the Club is reading. The update mask must be
+   *     associated with the key {@code "updateMask"} and have the value of a comma separated list
+   *     of the fields to be updated from the "club" key. To specify an update inside of the book
+   *     object inside of the Club, prepend the field name in the update mask with "currentBook.".
+   *     If no update mask exists, all fields from the "club" key will be used to update the Club
+   *     (and nested Book). If the request does not have either of these keys, or is syntactically
+   *     incorrect, the response object will send a "400 Bad Request error". If the user attempting
+   *     to update the Club is not the Club owner, the response object will send a "403 Forbidden
+   *     error"
+   * @param response the response from this method, will contain the updated Club in JSON format.
+   *     If the request object does not have a valid JSON body (as described in the request
+   *     parameter) this object will send a "400 Bad Request error". If the user attempting
+   *     to update the Club is not the Club owner, this object will send a "403 Forbidden error"
+   * @throws IOException if an input or output error is detected when the servlet handles the request
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Club club;
+    try {
+      JsonObject requestJson = gson.fromJson(request.getReader(), JsonObject.class);
+      JsonObject updatedClubJson = requestJson.getAsJsonObject(CLUB_FIELD_NAME);
+      JsonObject updatedBookJson = updatedClubJson.getAsJsonObject(Club.CURRENT_BOOK_FIELD_NAME);
+      JsonElement updateMask = requestJson.get(UPDATE_MASK_FIELD_NAME);
+      
+      // Get attempted updates for the Club and the Book inside the Club.
+      List<String> bookMask = new ArrayList<>();
+      List<String> clubMask = new ArrayList<>();
+      if (updateMask == null) {
+        bookMask = updateableBookFields;
+        clubMask = updateableClubFields;
+      } else {
+        // Cannot use stream here because Java will complain about local variables not being final.
+        String[] elements = updateMask.getAsString().split(",");
+        for (String element : elements) {
+          boolean isBookUpdate = element.contains(Club.CURRENT_BOOK_FIELD_NAME + ".");
```

Comment:
nit: More efficient and safer to use startsWith() https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#startsWith(java.lang.String)

Reply:
Good idea! This has been implemented and pushed.